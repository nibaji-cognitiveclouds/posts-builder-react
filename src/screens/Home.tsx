/** @format */

import {
	CircularProgress,
	Container,
	Table,
	TableBody,
	TableCell,
	TableContainer,
	TableFooter,
	TableHead,
	TableRow,
	Typography,
} from "@mui/material";
import axios from "axios";
import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

import { styles } from "../styles/screen";
import { newsResponse } from "../types/response";

const Home = () => {
	const [page, setPage] = useState(0);
	const [data, setData] = useState<newsResponse[]>([]);
	const [loading, setLoading] = useState(false);
	const [error, setError] = useState(false);

	const navigate = useNavigate();

	useEffect(() => {
		// getData();
		const inter = setInterval(() => {
			!loading &&
				setPage((prev) => {
					return ++prev;
				});
		}, 10000);
		return () => clearInterval(inter);
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, []);

	useEffect(() => {
		!loading && getData();
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [page]);

	function getData() {
		setLoading(true);
		axios
			.get(
				`https://hn.algolia.com/api/v1/search_by_date?query=story&page=${page}`
			)
			.then((res) => setData(Array.from(new Set([...data, ...res.data.hits]))))
			.catch((err) => setError(true))
			.finally(() => setLoading(false));
	}

	return (
		<Container style={styles.container}>
			<Typography variant="h4" marginY={3}>
				News
			</Typography>
			<TableContainer style={styles.tableContainer}>
				<Table stickyHeader={true}>
					<TableHead>
						<TableRow>
							<TableCell style={styles.tableHeaderCell}>Title</TableCell>
							<TableCell style={styles.tableHeaderCell}>Author</TableCell>
							<TableCell style={styles.tableHeaderCell}>Date</TableCell>
							<TableCell style={styles.tableHeaderCell}>URL</TableCell>
						</TableRow>
					</TableHead>
					<TableBody>
						{error ? (
							<Typography style={styles.error}>
								Something went wrong. Please try again!
							</Typography>
						) : (
							data?.map((item) => {
								const { author, created_at, title, url, objectID } = item;
								return (
									<TableRow
										key={objectID}
										onClick={() => {
											navigate(`/json`, {
												state: {
													item,
												},
											});
										}}
									>
										<TableCell style={styles.tableBodyCell}>
											{title || "-"}
										</TableCell>
										<TableCell style={styles.tableBodyCell}>
											{author || "-"}
										</TableCell>
										<TableCell style={styles.tableBodyCell}>
											{created_at || "-"}
										</TableCell>
										<TableCell style={styles.tableBodyCell}>
											{<a href={url}>{url}</a> || "-"}
										</TableCell>
									</TableRow>
								);
							})
						)}
					</TableBody>
					<TableFooter>
						<TableCell colSpan={4}>
							{loading && <CircularProgress style={styles.circularProgress} />}
						</TableCell>
					</TableFooter>
				</Table>
			</TableContainer>
		</Container>
	);
};

export default Home;
