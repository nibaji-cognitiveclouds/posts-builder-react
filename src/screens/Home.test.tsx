/** @format */

import { render, screen } from "@testing-library/react";
import Home from "./Home";

jest.mock("react-router-dom", () => {
	return {
		useNavigate: jest.fn(),
	};
});

describe("Home", () => {
	test("renders without crashing", () => {
		render(<Home />);
	});

	test("matches snapshot", () => {
		const view = render(<Home />);
		expect(view).toMatchSnapshot();
	});

	test("contains heading", () => {
		render(<Home />);
		const linkElement = screen.getByText("News");
		expect(linkElement).toBeInTheDocument();
	});

	test("has children", () => {
		const { container } = render(<Home />);
		const { children } = container;
		expect(children.length).toBeGreaterThan(0);
	});
});
