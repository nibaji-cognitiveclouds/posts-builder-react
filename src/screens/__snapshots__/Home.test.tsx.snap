// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
        style="display: flex; justify-content: center; align-items: center; flex-direction: column;"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 css-19v5luw-MuiTypography-root"
        >
          News
        </h4>
        <div
          class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
          style="border-radius: 20px; border-width: 10px; border-color: yellow;"
        >
          <table
            class="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
          >
            <thead
              class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                  scope="col"
                  style="background-color: rgb(139, 138, 37); color: white;"
                >
                  Title
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                  scope="col"
                  style="background-color: rgb(139, 138, 37); color: white;"
                >
                  Author
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                  scope="col"
                  style="background-color: rgb(139, 138, 37); color: white;"
                >
                  Date
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                  scope="col"
                  style="background-color: rgb(139, 138, 37); color: white;"
                >
                  URL
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
            />
            <tfoot
              class="MuiTableFooter-root css-9cgc89-MuiTableFooter-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-footer MuiTableCell-sizeMedium css-nfmglb-MuiTableCell-root"
                colspan="4"
              >
                <span
                  class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
                  role="progressbar"
                  style="width: 40px; height: 40px; margin-left: 409.6px; margin-top: 20px; margin-bottom: 20px;"
                >
                  <svg
                    class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                    viewBox="22 22 44 44"
                  >
                    <circle
                      class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                      cx="44"
                      cy="44"
                      fill="none"
                      r="20.2"
                      stroke-width="3.6"
                    />
                  </svg>
                </span>
              </td>
            </tfoot>
          </table>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      style="display: flex; justify-content: center; align-items: center; flex-direction: column;"
    >
      <h4
        class="MuiTypography-root MuiTypography-h4 css-19v5luw-MuiTypography-root"
      >
        News
      </h4>
      <div
        class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
        style="border-radius: 20px; border-width: 10px; border-color: yellow;"
      >
        <table
          class="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
        >
          <thead
            class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                scope="col"
                style="background-color: rgb(139, 138, 37); color: white;"
              >
                Title
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                scope="col"
                style="background-color: rgb(139, 138, 37); color: white;"
              >
                Author
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                scope="col"
                style="background-color: rgb(139, 138, 37); color: white;"
              >
                Date
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                scope="col"
                style="background-color: rgb(139, 138, 37); color: white;"
              >
                URL
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
          />
          <tfoot
            class="MuiTableFooter-root css-9cgc89-MuiTableFooter-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-footer MuiTableCell-sizeMedium css-nfmglb-MuiTableCell-root"
              colspan="4"
            >
              <span
                class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
                role="progressbar"
                style="width: 40px; height: 40px; margin-left: 409.6px; margin-top: 20px; margin-bottom: 20px;"
              >
                <svg
                  class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                  viewBox="22 22 44 44"
                >
                  <circle
                    class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                    cx="44"
                    cy="44"
                    fill="none"
                    r="20.2"
                    stroke-width="3.6"
                  />
                </svg>
              </span>
            </td>
          </tfoot>
        </table>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
